pallets/subtensor/src/subnets/weights.rs

```diff
@@ -41,21 +41,32 @@ impl<T: Config> Pallet<T> {
             Error::<T>::WeightsCommitNotAllowed
         );
 
-        // Insert the new commit with the current block number.
+        // Insert the new commit with the current block number
         let current_block = Self::get_current_block_as_u64();
         WeightCommits::<T>::insert(netuid, &who, (commit_hash, current_block));
 
-        // Retain only commits where committed_block < current_block - commit_reveal_interval * 2
+        // Calculate the threshold block beyond which old commits should be removed
         let interval = Self::get_commit_reveal_weights_interval(netuid).saturating_mul(2);
         let threshold_block = current_block.saturating_sub(interval);
 
-        // Iterate through the commits and remove old entries
-        WeightCommits::<T>::drain_prefix(netuid).for_each(|(who, (_hash, commit_block))| {
-            if commit_block < threshold_block {
-                WeightCommits::<T>::remove(netuid, who);
-            }
-        });
-
+        // Retrieve all commits for the given network as a vector, sorted by block number
+        let mut commits: Vec<(T::AccountId, (H256, u64))> =
+            WeightCommits::<T>::iter_prefix(netuid).collect();
+
+        // Use binary search to find the first valid commit that is above the threshold
+        let first_valid_index = commits
+            .binary_search_by(|&(_, (_, commit_block))| commit_block.cmp(&threshold_block))
+            .unwrap_or_else(|index| index); // If not found, return the position of the first valid commit
+
+        // If there are old commits, remove them in bulk
+        if first_valid_index > 0 {
+            // Remove old commits that are below the threshold
+            commits
+                .drain(0..first_valid_index)
+                .for_each(|(account, _)| {
+                    WeightCommits::<T>::remove(netuid, &account);
+                });
```

Comment:
what's `account` in this context?

Doesn't this call `WeightCommits::<T>::remove();` repeatedely, where each of those operations rewrites the entire list in memory O(N*M), where it could get a range from `first_valid_index..end` and replace the substructure with it in one memcopy O(M)?